scripts/redcap/wrong_date_associations.py

```diff
@@ -148,7 +150,41 @@ def log_row(row: pd.Series, **kwargs):
     for _, row in errors_df.reset_index().iterrows():
         log_row(row, uid=uid_template, **kwargs)
 
+def get_special_cases(marks, session):
+    """
+    Looks at special_cases.yml file, finds the special cases under the file name section,
+    and then excludes any cases that match the special cases list.
+
+    Note (for Chris use):
+    - 'redcap_event_name' is arm name
+    - 'study_id' inclues up to letter and number
+    - 'visit_date' is looking at date --> not correct!
 
+    Each row in the dataframe has:
+    row.name -> [0] is the subject ID, [1] is the arm name
+    """
+
+    # Connect to the special_cases.yml file
+    sibis_config = session.get_operations_dir()
+    special_cases_file = os.path.join(sibis_config, 'special_cases.yml')
+    if not os.path.isfile(special_cases_file):
+        slog.info("wrong_date_associations","Error: The following file does not exist : " + special_cases_file)
+        sys.exit(1)
+
+    # Get a list of the specific cases that should be inspected
+    with open(special_cases_file) as fi:
+        special_cases = yaml.load(fi)
+        exceptions_data = special_cases.get('wrong_date_associations', {})
+
+    # Loop through each exception and drop the corresponding row
+    for exception in exceptions_data:
+        for index, row in marks.iterrows():
```

Comment:
Quick attempt to optimize this:

```python
marks.set_index('form_date_var', append=True, inplace=True)
invalid_marks_idx = marks.index.isin(exceptions_data)  # might require some reshaping of exceptions_data - assuming List[Tuple[str, str, str]]?
invalid_marks = marks.loc[invalid_marks_idx]  # and determine from this if any exceptions_data is without a matching mark
marks = marks.loc[~invalid_marks_idx]

return marks
```

But if you can't make this work, we can go with this - we have little enough data that we can afford a little O(n^2), as a treat.

Reply:
Can definitely try and optimize this after our meeting, while I'm concurrently working on another script? I think that could be helpful, especially since a verbal explanation of the code could be great!